### PR TITLE
SoC14 wxWebViewChromium merge

### PR DESCRIPTION
**Pull request by sjlamerton**
First attempt at a merge of the 2014 wxWebViewChromium Summer of Code work. Straight rebase of the Haojian's changes onto wxWidgets master. Conflicting commits were skipped as they related to build files that were later removed. One additional commit to fix the msw build files after their recent changes.

### CODE REVIEW

**Comment by steve-lamerton** [comment]
@vadz I rebased this earlier today with a few changes since the last time we discussed this, any chance you could take a look again?

File: configure.in [hunk 1]
```diff
@@ -8233,7 +8239,7 @@ echo "                                       zlib               ${wxUSE_ZLIB}"
 echo "                                       expat              ${wxUSE_EXPAT}"
 echo "                                       libmspack          ${wxUSE_LIBMSPACK}"
 echo "                                       sdl                ${wxUSE_LIBSDL}"
-
+echo "                                       webview_chromium   ${wxUSE_WEBVIEW_CHROMIUM}"
```
Comment (vadz) [inline configure.in]: Ideal would be to have something like

    webview         yes (with the following backends: WebKit Chromium)

to provide maximum of information on a single line.

File: configure.in [hunk 2]
```diff
@@ -7047,6 +7048,11 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
     USE_WEBVIEW=1
     AC_DEFINE(wxUSE_WEBVIEW)
     SAMPLES_SUBDIRS="$SAMPLES_SUBDIRS webview"
+    if test "$wxUSE_WEBVIEW_CHROMIUM" = "yes"; then
+        USE_WEBVIEW_CHROMIUM=1
```
Comment (vadz) [inline configure.in]: This doesn't seem right, the option is "yes" by default (not sure if it should be, if not, `DEFAULT_wxUSE_WEBVIEW_CHROMIUM` should be set to "no") and we don't check that we have the required headers/libraries here, so it looks like the compilation will fail out of the box.

We should have `AC_TRY_LINK` check here.

File: include/wx/webview_chromium.h [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+#include "wx/webview.h"
+#include "wx/sharedptr.h"
+#include "wx/vector.h"
+
+#ifdef __VISUALC__
+#pragma warning(push)
+#pragma warning(disable:4100)
+#endif
+
+#include "include/cef_browser.h"
```
Comment (vadz) [inline include/wx/webview_chromium.h]: This would probably represent a significant change, but it would be great if we could avoid including all this CEF stuff from this public header. AFAICS the only thing actually needed here is `m_clientHandler` and we could just wrap it into some `wxWebViewChromiumImpl` to avoid having it here.

There would be quite a few advantages of keeping this header free of CEF stuff, in particular we could even allow people to use this backend without installing CEF headers at all if we just build the official binaries with it, it would be enough to distribute just the DLLs.

File: interface/wx/webview_chromium.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+    - C/C++ - Language - Enable Run-Time Type Information - Yes
+
+    The following release branches of CEF have been tested: 1650, 1750, 1916,
+    2062 and 2171.
+
+    @section instructions Build Instructions
+
+    __General__
+
+    The wxWebViewChromium backend is built into a seperate webview_chromium
```
Comment (vadz) [inline interface/wx/webview_chromium.h]: Typo s/seperate/separate/

File: interface/wx/webview_chromium.h [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+    library which depends on the webview library.
+
+    Once you have a copy of CEF3, either from compiling it yourself or using
+    prebuild binaries you should copy it into `wx_root/src/cef`. To run the
```
Comment (vadz) [inline interface/wx/webview_chromium.h]: s/prebuild/prebuilt/

File: samples/webview/Info_cocoa.plist [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	<string>6.0</string>
+	<key>CFBundleLocalizations</key>
+	<array>
+		<string>de</string>
+		<string>en</string>
+		<string>fr</string>
+		<string>it</string>
+	</array>
+	<key>CFBundleLongVersionString</key>
+	<string>3.1.0, (c) 2005-2013 wxWidgets</string>
```
Comment (vadz) [inline samples/webview/Info_cocoa.plist]: Very minor, but copyright year is off by 2 by now...

(here and below)

File: src/common/webview_chromium.cpp [hunk 7]
```diff
@@ -0,0 +1,809 @@
...
+#endif
+
+extern WXDLLIMPEXP_DATA_WEBVIEW_CHROMIUM(const char) wxWebViewBackendChromium[] = "wxWebViewChromium";
+
+wxIMPLEMENT_DYNAMIC_CLASS(wxWebViewChromium, wxWebView);
+
+class wxStringVisitor : public CefStringVisitor
+{
+public:
+    enum StringType {
```
Comment (vadz) [inline src/common/webview_chromium.cpp]: Minor: `{` should be on its own line.

File: src/common/webview_chromium.cpp [hunk 8]
```diff
@@ -0,0 +1,809 @@
...
+
+bool wxWebViewChromium::Create(wxWindow* parent,
+           wxWindowID id,
+           const wxString& url,
+           const wxPoint& pos,
+           const wxSize& size,
+           long style,
+           const wxString& name)
+{
+#ifdef __WXMSW__
```
Comment (vadz) [inline src/common/webview_chromium.cpp]: What's the difference between the MSW and OSX case? I.e. why don't we check for the return value of `Create()` in the latter and why don't we just have `#ifdef __WXGTK__` and `#else` branches instead of checking for each platform?

File: src/common/webview_chromium.cpp [hunk 9]
```diff
@@ -0,0 +1,809 @@
...
+    if ( !wxControl::Create(parent, id, pos, size, style,
+                           wxDefaultValidator, name) )
+    {
+        return false;
+    }
+#endif
+
+#ifdef __WXGTK__
```
Comment (vadz) [inline src/common/webview_chromium.cpp]: A comment explaining why is this needed and, in particular, why can't we just call base class `Create()` with `style | wxHSCROLL | wxVSCROLL` would be helpful.

File: src/common/webview_chromium.cpp [hunk 10]
```diff
@@ -0,0 +1,809 @@
...
+                                int argc, char* argv[])
+#endif
+{
+#ifdef __WXMSW__
+    CefMainArgs args(wxGetInstance());
+#else
+    CefMainArgs args(argc, argv);
+#endif
+    // If there is no subprocess then we need to execute on this process
+    if ( path == "" )
```
Comment (vadz) [inline src/common/webview_chromium.cpp]: Minor: should be `path.empty()`

File: src/common/webview_chromium.cpp [hunk 11]
```diff
@@ -0,0 +1,809 @@
...
+        if ( code >= 0 )
+            return false;
+    }
+
+    CefSettings settings;
+#if CHROME_VERSION_BUILD >= 1750
+    settings.no_sandbox = true;
+#endif
+
+#ifdef __WXDEBUG__
```
Comment (vadz) [inline src/common/webview_chromium.cpp]: This is not appropriate, `__WXDEBUG__` is defined even in release build by default in wx3.

It would probably be better to either use some `wxWEBVIEW_DEBUG` environment variable or provide a method for enabling debug logging.

**Comment by vadz** [comment]
Thanks again for your work, I can't really comment on the interesting parts, i.e. the code actually using CEF API, but globally it seems good except for the few things pointed out in the comments.

I also wonder why do we need a separate sample for this, could we just allow using this backend as an option in the normal webview sample? I'd really like to have just one sample instead of two similar but different ones.

**Comment by steve-lamerton** [comment]
I have made most of the fixes, the Impl changes would be quite big but I'll take a look later this week, I'll also update the debug output then. As for the samples I think the reason that there are two is because it isn't possible to compile the webview_chromium library on Windows if CEF is not present which would limit the number of people who could build the sample as it would fail to link.

**Comment by vadz** [comment]
About the sample: could we at least reuse the sources? I.e. we could still have a different directory with its own makefiles, but just `#include "../webview/webview.cpp"` from `webview_chromium.cpp` or something like this. Duplicating so much code really doesn't sit well with me.

**Comment by mths0x5f** [comment]
Sorry for the intromission, but how is this going? What can someone do to help?

**Comment by steve-lamerton** [comment]
The only item left to resolve before the merge was the duplicated code in the samples, it wouldn't take much to resolve that and then I think we should be able to close this pull request.

**Comment by mths0x5f** [comment]
Well... I do not know if it helps, but I have edited some code and made a pull request to you, @steve-lamerton. I'm sorry if inconvenient.

https://github.com/steve-lamerton/wxWidgets/pull/1

**Comment by steve-lamerton** [comment]
Thanks, I will take a look later this week.

**Comment by tierra** [comment]
It's probably too late to look into getting this into 3.1.0, but it would still be great if @sjlamerton or @hokein could look at rebasing this, resolving conflicts, and testing this out on latest master. I'd still love to see this make it into a future 3.1.x release even if it's not 3.1.0.

**Comment by mtangoo** [comment]
I would also love to see this merged. It will relieve developers relying on web a lot of pain especially with versions of Native engines available (some are still using very old MS browsers in their systems)

**Comment by vadz** [comment]
I believe this is obsoleted by #706, please let me know if I was wrong to close this one.
